jersey/src/main/java/org/pac4j/jax/rs/jersey/features/Pac4JValueFactoryProvider.java

```diff
@@ -165,69 +159,94 @@ public Binder() {
-            }).in(Singleton.class);
+            bind(ProfileInjectionResolver.class)
+                .to(new GenericType<InjectionResolver<Pac4JProfile>>(){})
+                .in(Singleton.class);
+
+            bind(ProfileManagerInjectionResolver.class)
+                .to(new GenericType<InjectionResolver<Pac4JProfileManager>>(){})
+                .in(Singleton.class);
         }
     }
 
-    static class ProfileManagerValueFactory extends AbstractContainerRequestValueFactory<ProfileManager<CommonProfile>>
-            implements ProfileManagerFactory {
-
+    static class ProfileManagerValueFactory implements ProfileManagerFactory{
         @Context
-        Providers providers;
+        private Providers providers;
+
+        ProfileManagerValueFactory(Providers providers) {
+            this.providers = providers;
+        }
+
+        @Override
+        public ProfileManager<CommonProfile> apply(ContainerRequest containerRequest) {
+            return new RequestProfileManager(new RequestJaxRsContext(providers, containerRequest))
+                .profileManager();
+        }
+    }
 
+    static class ProfileValueFactory implements ProfileFactory {
         @Override
-        public ProfileManager<CommonProfile> provide() {
-            return new RequestProfileManager(new RequestJaxRsContext(providers, getContainerRequest()))
-                    .profileManager();
+        public CommonProfile apply(ContainerRequest containerRequest) {
+            return optionalProfile(containerRequest)
+                .orElseThrow(() -> {
+                    LOG.debug("Cannot inject a Pac4j profile into an unauthenticated request, responding with 401");
+                    return new WebApplicationException(401);
+                });
         }
     }
 
-    static class ProfileValueFactory extends AbstractContainerRequestValueFactory<CommonProfile>
-            implements ProfileFactory {
+    static class OptionalProfileValueFactory implements OptionalProfileFactory {
         @Override
-        public CommonProfile provide() {
-            return new RequestCommonProfile(new RequestPac4JSecurityContext(getContainerRequest())).profile()
-                    .orElseThrow(() -> {
-                        LOG.debug("Cannot inject a Pac4j profile into an unauthenticated request, responding with 401");
-                        return new WebApplicationException(401);
-                    });
+        public Optional<CommonProfile> apply(ContainerRequest containerRequest) {
+            return optionalProfile(containerRequest);
         }
     }
 
-    static class OptionalProfileValueFactory extends AbstractContainerRequestValueFactory<Optional<CommonProfile>>
-            implements OptionalProfileFactory {
+    private static Optional<CommonProfile> optionalProfile(ContainerRequest containerRequest) {
+        RequestPac4JSecurityContext securityContext = new RequestPac4JSecurityContext(containerRequest.getSecurityContext());
```

Comment:
I think you should be able to write: `new RequestPac4JSecurityContext(containerRequest)`

Reply:
Right, thanks!